keps/sig-network/1880-multiple-service-cidrs/README.md

```diff
@@ -0,0 +1,954 @@
+
+- users can not resize or increase the ranges of IPs assigned to Services, causing problems when
+  there are overlapping networks or the cluster run out of available IPs.
+- the Service IP range is not exposed, so it is not possible for other components in the cluster to
+  consume it
+- the configuration is per node, there is no consensus and different apiservers can fight and delete
+  others IPs.
+- the allocation logic is racy, with the possibility of leaking ClusterIPs
+  <https://github.com/kubernetes/kubernetes/issues/87603>
+- the size of the Service Cluster CIDR, for IPv4 the prefix size is limited to /12, however, for
+  IPv6 it is limited to /112 or fewer. This restriction is causing issues for IPv6 users, since /64
+  is the standard and minimum recommended prefix length
+- only is possible to use reserved a range of Service IP addresses using the feature gate
+  "keps/sig-network/3070-reserved-service-ip-range"
+  <https://github.com/kubernetes/kubernetes/issues/95570>
+
+### Goals
+
+Implement a new allocation logic for Services IPs that:
+
+- scales well with the number of reservations
+- the number of allocation is tunable
+- is completely backwards compatible
+
+### Non-Goals
+
+- Any change unrelated to Services
+- Any generalization of the API model that can evolve onto something different, like an IPAM API, or
+  collide with existing APIs like Ingress and GatewayAPI.
+- NodePorts use the same allocation model than ClusterIPs, but it is totally out of scope of this
+  KEP. However, this KEP can be a proof that a similar model will work for NodePorts too.
+- Change the default IPFamily used for Services IP allocation, the defaulting depends on the
+  services.spec.IPFamily and services.spec.IPFamilyPolicy, a simple webhook or an admission plugin
+  can set this fields to the desired default, so the allocation logic doesn't have to handle it.
+- Removing the apiserver flags that define the service IP CIDRs, though that may be possible in the future.
+
+## Proposal
+
+The proposal is to implement a new allocator logic that uses 2 new API Objects: ServiceCIDRConfig
+and IPAddress, and allow users to dynamically increase the number of Services IPs available by
+creating new ServiceCIDRConfigs.
+
+The new allocator will be able to "automagically" consume IPs from any ServiceCIDRConfig available, we can
+think about this model, as the same as adding more disks to a Storage system to increase the
+capacity.
+
+To simplify the model, make it backwards compatible and to avoid that it can evolve into something
+different and collide with other APIs, like Gateway APIs, we are adding the following constraints:
+
+- a ServiceCIDRConfig will be immutable after creation (to be revisited before Beta).
+- a ServiceCIDRConfig can only be deleted if there are no Service IP associated to it (enforced by finalizer).
+- there can be overlapping ServiceCIDRConfigs.
```

Comment:
How does this work? If there are overlapping ranges couldn't we get duplicate IPs? Or is it not true that there is a 1-N relationship between `ServiceCIDRConfig` and `IPAddress`, but rather N-N?

Reply:
There can not be duplicate IPAddresses, it uses the `Name` for storing the IP in canonical format, and names are unique, so IPAddress are unique ... the allocator has to handle the overlapping, just increase or shrink the available ranges they can peek addresses

Reply:
The CIDRs are merged in memory - an IP doesn't come from a specific range object, but "any range that includes that IP"

So for example, if we had a range for 10.0.0.0/24 and 10.0.0.0/20, and a service was using 10.0.0.10 - either range satisifes, and the other could be deleted, bu 10.0.10.0 is only satisifed by the larger /20

Reply:
So, the statement that `ServiceCIDRConfig -> IPAddress` is `1 -> N` is not true. It is `N -> N`. Or is `->` some other relation than "IP in CIDR"? This is important because later I believe you rely on this statement being true.

Reply:
added a comment that hopes it clarifies it, the relation 1 -> N is not really ServiceCIDRConfig -> IPAddress is ServiceCIDR defined by ServiceCIDRConfig (it can be multiples) ->IPAddress